### DISCUSSION TITLE
[Code Snippet] Preview window with updating plot

### DISCUSSION THREAD

**Discussion by LangLangBart** [junegunn/fzf#3686] [General]
I'd like to share these code snippets demonstrating the use of **asciigraph** and **diagram** to
draw real-time plot in the `fzf` preview-window. 

These demos use the `ps` command to monitor the CPU usage of the `fzf` build process.

The graph updates every 0.5 seconds until the process ends. When a process ends, the list updates
using the `--listen`[^1] flag and the `FZF_PORT`[^2] environment variable.

---

**asciigraph** - https://github.com/guptarohit/asciigraph

<img src="https://github.com/junegunn/fzf/assets/92653266/49251576-297f-4358-84bc-389b6271d123" width="600">

```bash
ps -u "$(id -u)" -o pid,pcpu,pmem,ucomm,command -r |
	fzf --prompt "Current processes [$(id -un)] > " \
		--header-lines 1 \
		--listen \
		--nth '4,5,1' \
		--reverse \
		--preview $'
			while kill -0 $PPID;do
				if kill -0 {1} 2>/dev/null; then
					sleep 0.5
					ps -p {1} -o %cpu= 2>/dev/null
				else
					printf "\033[2J" >&2
					echo "{4} with PID {1} has ended" >&2
					echo "Reloading the list..." >&2
					sleep 1
					curl -sX POST  "localhost:$FZF_PORT" --data "reload:ps -u $UID -o pid,pcpu,pmem,ucomm,command -r"
					break
				fi
			done | asciigraph -r -h 10 -w 40 -c "%CPU usage {4}" -sc yellow'
```

---

**diagram** - https://github.com/tehmaze/diagram

<img src="https://github.com/junegunn/fzf/assets/92653266/e7aa2840-5bd7-42fb-8a2c-9bf542c7d7f5" width="600">

```bash
ps -u "$(id -u)" -o pid,pcpu,pmem,ucomm,command -r |
	fzf --prompt "Current processes [$(id -un)] > " \
		--header-lines 1 \
		--listen \
		--nth '4,5,1' \
		--reverse \
		--preview $'
			while kill -0 $PPID;do
				if kill -0 {1} 2>/dev/null; then
					sleep 0.5
					printf "\033[2J" >&2
					ps -p {1} -o %cpu= 2>/dev/null
				else
					printf "\033[2J" >&2
					echo "{4} with PID {1} has ended" >&2
					echo "Reloading the list..." >&2
					sleep 1
					curl -sX POST  "localhost:$FZF_PORT" --data "reload:ps -u $UID -o pid,pcpu,pmem,ucomm,command -r"
					break
				fi
			done | diagram --batch --height 10 --width 40'
```

[^1]: [0.36.0 CHANGELOG.md](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0360)
[^2]: [0.39.0 CHANGELOG.md](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0390)

**Comment by junegunn** [comment]
Beautiful, thanks for sharing!

**Comment by damnskippy** [comment]
wow!
